app/javascript/mastodon/features/ui/components/tabs_bar.js

```diff
@@ -25,16 +26,46 @@ export function getLink (index) {
 @injectIntl
 export default class TabsBar extends React.Component {
 
+  static contextTypes = {
+    router: PropTypes.object.isRequired,
+  }
+
   static propTypes = {
     intl: PropTypes.object.isRequired,
   }
 
+  setRef = ref => {
+    this.node = ref;
+  }
+
+  handleClick = (e) => {
+    e.preventDefault();
+    e.persist();
+
+    requestAnimationFrame(() => {
+      const currentTab = this.node.querySelector('.tabs-bar__link.active');
+      const nextTab = e.nativeEvent.path.find(node => node.classList.contains('tabs-bar__link'));
+      const { props: { to } } = links[Array(...this.node.childNodes).indexOf(nextTab)];
+
+      currentTab.classList.remove('active');
+
+      const listener = debounce(() => {
+        nextTab.removeEventListener('transitionend', listener);
+        this.context.router.history.push(to);
+      }, 50);
```

Comment:
What's with the debounce? `transitionend` should only be called once, exactly when the transition has ended.